core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java

```diff
@@ -280,6 +280,13 @@ protected boolean needToCast(SqlValidatorScope scope, SqlNode node,
       return false;
     }
 
+    if (node.getKind() == SqlKind.LITERAL
```

Comment:
I'm not particularly familiar with the parts of validator. Thanks for any advice. 
Here's what I understand:
In [CALCITE-5156], this part of the code is removed here to implement the implicit conversion of In. 
```  
 if (romType.getPrecedenceList().containsType(toType)
        && SqlTypeUtil.isIntType(fromType)
        && SqlTypeUtil.isIntType(toType)) {
      return false;
    }
```
But for Literal, In will be converted to Value at ToRel, Value allows the type of Literal in Tuple to be not exactly the same as the type of Value (because for Literal, the type often depends on the external call). So the type takes precedence over Value. Therefore, the conversion of Literal is not necessary here, and is described in the code comment as follows
```
private RexLiteral convertLiteral(SqlLiteral sqlLiteral,
    Blackboard bb, RelDataType type) {
  RexNode literalExpr = exprConverter.convertLiteral(bb, sqlLiteral);

  if (!(literalExpr instanceof RexLiteral)) {
    assert literalExpr.isA(SqlKind.CAST);
    RexNode child = ((RexCall) literalExpr).getOperands().get(0);
    assert RexLiteral.isNullLiteral(child);

    // NOTE jvs 22-Nov-2006:  we preserve type info
    // in LogicalValues digest, so it's OK to lose it here
    return (RexLiteral) child;
  }
}
```
And Literal will not be inferred as a smallint. 
```
  @Override public RelDataType createSqlType(RelDataTypeFactory typeFactory) {
    if (exact) {
      int scaleValue = requireNonNull(scale, "scale");
      if (0 == scaleValue) {
        try {
          BigDecimal bd = getValueNonNull();
          SqlTypeName result;
          // Will throw if the number cannot be represented as a long.
          long l = bd.longValueExact();
          if ((l >= Integer.MIN_VALUE) && (l <= Integer.MAX_VALUE)) {
            result = SqlTypeName.INTEGER;
          } else {
            result = SqlTypeName.BIGINT;
          }
          return typeFactory.createSqlType(result);
        } catch (ArithmeticException ex) {
          // This indicates that the value does not fit in any integer type.
          // Fallback to DECIMAL.
        }
      }
      // else we have a decimal
      return typeFactory.createSqlType(
          SqlTypeName.DECIMAL,
          requireNonNull(prec, "prec"),
          scaleValue);
    }
```
Therefore, for an integer number that is always < Interger.MAX_VALUE, CAST is always added here if the function was originally inferred to not be an Integer.  This leads to the Bad Plan described in Jira.

Reply:
Is the plan bad because it's incorrect, or because it's not efficient?
I think correctness is not negotiable.
Efficiency can be improved, but never at the cost of correctness.
One rule for correctness of a plan is that it typechecks.

In general, you should not remove a cast that may change the semantics of the program.
A cast can, for example, cause a runtime exception, if the value does not fit in the target type. It's not obvious to me that the cast you are removing here is not of this kind.

The semantics of the literal is reflected not only in the value of this particular literal, but also in its type, which may be propagated to other expressions that make use of this literal.

Regarding the union of values, I have a PR in the works for https://issues.apache.org/jira/browse/CALCITE-6350, which will insert even *more* casts.

Reply:
Thanks for the review, I agree that the cast prior to this pr is correct. The problem I'm trying to solve doesn't conflict with this, I'll close this PR and re-edit the issue.
BTW, for the raggedUnionTypesToVarying setting in your PR, that's what I've been puzzled about as well. And for the CHAR type cast, I once mentioned an ISSUE for that as well, [CALCITE-6520], and I'm not sure if your PR addresses that as well.